File: examples/mcp-clients/python/simple_client.py
#!/usr/bin/env python3
"""
Simple Python MCP Client for RAG MCP Server

This script demonstrates how to connect to and interact with the RAG MCP Server
using HTTP transport.
"""

import requests
import json
import sys
import time
from typing import Dict, Any, Optional


class RAGMCPClient:
    """Simple MCP client for the RAG MCP Server."""
    
    def __init__(self, base_url: str = "http://localhost:8000"):
        """
        Initialize the MCP client.
        
        Args:
            base_url: Base URL of the RAG MCP Server
        """
        self.base_url = base_url.rstrip('/')
        self.mcp_url = f"{self.base_url}/mcp/"
        self.health_url = f"{self.base_url}/health"
        self.request_id = 1
        self.timeout = 30
    
    def _make_request(self, method: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """
        Make an MCP JSON-RPC request.
        
        Args:
            method: MCP method name
            params: Method parameters
            
        Returns:
            JSON-RPC response
            
        Raises:
            Exception: If request fails
        """
        payload = {
            "jsonrpc": "2.0",
            "id": self.request_id,
            "method": method,
            "params": params
        }
        self.request_id += 1
        
        try:
            response = requests.post(
                self.mcp_url,
                json=payload,
                headers={"Content-Type": "application/json"},
                timeout=self.timeout
            )
            response.raise_for_status()
            
            result = response.json()
            
            # Check for JSON-RPC errors
            if "error" in result:
                error = result["error"]
                raise Exception(f"MCP Error {error.get('code', 'UNKNOWN')}: {error.get('message', 'Unknown error')}")
            
            return result
            
        except requests.exceptions.RequestException as e:
            raise Exception(f"HTTP request failed: {e}")
        except json.JSONDecodeError as e:
            raise Exception(f"Invalid JSON response: {e}")
    
    def get_server_status(self) -> Dict[str, Any]:
        """
        Get server status.
        
        Returns:
            Server status information
        """
        response = self._make_request("tools/call", {
            "name": "get_server_status",
            "arguments": {}
        })
        return response.get("result", {})
    
    def ingest_csv(self, file_path: str) -> Dict[str, Any]:
        """
        Ingest a CSV file.
        
        Args:
            file_path: Path to the CSV file
            
        Returns:
            Ingestion result
        """
        response = self._make_request("tools/call", {
            "name": "ingest_csv",
            "arguments": {"file_path": file_path}
        })
        return response.get("result", {})
    
    def query_documents(self, query: str, max_results: int = 10, 
                       score_threshold: Optional[float] = None) -> Dict[str, Any]:
        """
        Query documents.
        
        Args:
            query: Natural language query
            max_results: Maximum number of results
            score_threshold: Minimum similarity score
            
        Returns:
            Query results
        """
        args = {"query": query, "max_results": max_results}
        if score_threshold is not None:
            args["score_threshold"] = score_threshold
        
        response = self._make_request("tools/call", {
            "name": "query_documents",
            "arguments": args
        })
        return response.get("result", [])
    
    def get_health(self) -> Dict[str, Any]:
        """
        Get health status via HTTP endpoint.
        
        Returns:
            Health status information
        """
        try:
            response = requests.get(self.health_url, timeout=10)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            raise Exception(f"Health check failed: {e}")
    
    def test_connection(self) -> bool:
        """
        Test connection to the server.
        
        Returns:
            True if connection successful, False otherwise
        """
        try:
            # Test health endpoint
            health = self.get_health()
            print(f"✓ Health check passed: {health.get('status')}")
            
            # Test MCP endpoint
            status = self.get_server_status()
            print(f"✓ MCP connection successful: {status.get('server_name')} v{status.get('version')}")
            print(f"  Status: {status.get('status')}")
            print(f"  Components: {status.get('components')}")
            
            return True
            
        except Exception as e:
            print(f"✗ Connection test failed: {e}")
            return False


def main():
    """Main function demonstrating client usage."""
    
    # Parse command line arguments
    if len(sys.argv) > 1:
        server_url = sys.argv[1]
    else:
        server_url = "http://localhost:8000"
    
    print(f"Connecting to RAG MCP Server at: {server_url}")
    print("=" * 50)
    
    # Create client
    client = RAGMCPClient(server_url)
    
    # Test connection
    if not client.test_connection():
        print("\nConnection test failed. Please check:")
        print("1. Server is running with HTTP transport enabled")
        print("2. Server URL is correct")
        print("3. Network connectivity")
        sys.exit(1)
    
    print("\n" + "=" * 50)
    print("Connection successful! You can now use the client.")
    print("=" * 50)
    
    # Interactive mode
    while True:
        print("\nAvailable commands:")
        print("1. status - Get server status")
        print("2. health - Get health status")
        print("3. ingest <file_path> - Ingest CSV file")
        print("4. query <query_text> - Query documents")
        print("5. quit - Exit")
        
        try:
            command = input("\nEnter command: ").strip().split()
            
            if not command:
                continue
            
            cmd = command[0].lower()
            
            if cmd == "quit" or cmd == "exit":
                break
            elif cmd == "status":
                status = client.get_server_status()
                print(json.dumps(status, indent=2))
            elif cmd == "health":
                health = client.get_health()
                print(json.dumps(health, indent=2))
            elif cmd == "ingest":
                if len(command) < 2:
                    print("Usage: ingest <file_path>")
                    continue
                file_path = " ".join(command[1:])
                print(f"Ingesting CSV file: {file_path}")
                result = client.ingest_csv(file_path)
                print(json.dumps(result, indent=2))
            elif cmd == "query":
                if len(command) < 2:
                    print("Usage: query <query_text>")
                    continue
                query_text = " ".join(command[1:])
                print(f"Querying documents: {query_text}")
                results = client.query_documents(query_text)
                print(json.dumps(results, indent=2))
            else:
                print(f"Unknown command: {cmd}")
                
        except KeyboardInterrupt:
            print("\nExiting...")
            break
        except Exception as e:
            print(f"Error: {e}")


if __name__ == "__main__":
    main()

File: examples/mcp-clients/nodejs/simple-client.js
#!/usr/bin/env node

/**
 * Simple Node.js MCP Client for RAG MCP Server
 * 
 * This script demonstrates how to connect to and interact with the RAG MCP Server
 * using HTTP transport from Node.js applications.
 */

const axios = require('axios');
const readline = require('readline');

class RAGMCPClient {
    /**
     * Initialize the MCP client.
     * 
     * @param {string} baseUrl - Base URL of the RAG MCP Server
     */
    constructor(baseUrl = 'http://localhost:8000') {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.mcpUrl = `${this.baseUrl}/mcp/`;
        this.healthUrl = `${this.baseUrl}/health`;
        this.requestId = 1;
        this.timeout = 30000; // 30 seconds
    }

    /**
     * Make an MCP JSON-RPC request.
     * 
     * @param {string} method - MCP method name
     * @param {Object} params - Method parameters
     * @returns {Promise<Object>} JSON-RPC response
     */
    async makeRequest(method, params) {
        const payload = {
            jsonrpc: '2.0',
            id: this.requestId++,
            method: method,
            params: params
        };

        try {
            const response = await axios.post(this.mcpUrl, payload, {
                headers: { 'Content-Type': 'application/json' },
                timeout: this.timeout
            });

            const result = response.data;

            // Check for JSON-RPC errors
            if (result.error) {
                const error = result.error;
                throw new Error(`MCP Error ${error.code || 'UNKNOWN'}: ${error.message || 'Unknown error'}`);
            }

            return result;

        } catch (error) {
            if (error.response) {
                throw new Error(`HTTP ${error.response.status}: ${error.response.statusText}`);
            } else if (error.request) {
                throw new Error(`Network error: ${error.message}`);
            } else {
                throw new Error(`Request error: ${error.message}`);
            }
        }
    }

    /**
     * Get server status.
     * 
     * @returns {Promise<Object>} Server status information
     */
    async getServerStatus() {
        const response = await this.makeRequest('tools/call', {
            name: 'get_server_status',
            arguments: {}
        });
        return response.result || {};
    }

    /**
     * Ingest a CSV file.
     * 
     * @param {string} filePath - Path to the CSV file
     * @returns {Promise<Object>} Ingestion result
     */
    async ingestCsv(filePath) {
        const response = await this.makeRequest('tools/call', {
            name: 'ingest_csv',
            arguments: { file_path: filePath }
        });
        return response.result || {};
    }

    /**
     * Query documents.
     * 
     * @param {string} query - Natural language query
     * @param {number} maxResults - Maximum number of results
     * @param {number|null} scoreThreshold - Minimum similarity score
     * @returns {Promise<Array>} Query results
     */
    async queryDocuments(query, maxResults = 10, scoreThreshold = null) {
        const args = { query, max_results: maxResults };
        if (scoreThreshold !== null) {
            args.score_threshold = scoreThreshold;
        }

        const response = await this.makeRequest('tools/call', {
            name: 'query_documents',
            arguments: args
        });
        return response.result || [];
    }

    /**
     * Get health status via HTTP endpoint.
     * 
     * @returns {Promise<Object>} Health status information
     */
    async getHealth() {
        try {
            const response = await axios.get(this.healthUrl, { timeout: 10000 });
            return response.data;
        } catch (error) {
            if (error.response) {
                throw new Error(`Health check failed: HTTP ${error.response.status}`);
            } else {
                throw new Error(`Health check failed: ${error.message}`);
            }
        }
    }

    /**
     * Test connection to the server.
     * 
     * @returns {Promise<boolean>} True if connection successful
     */
    async testConnection() {
        try {
            // Test health endpoint
            const health = await this.getHealth();
            console.log(`✓ Health check passed: ${health.status}`);

            // Test MCP endpoint
            const status = await this.getServerStatus();
            console.log(`✓ MCP connection successful: ${status.server_name} v${status.version}`);
            console.log(`  Status: ${status.status}`);
            console.log(`  Components:`, status.components);

            return true;

        } catch (error) {
            console.error(`✗ Connection test failed: ${error.message}`);
            return false;
        }
    }
}

/**
 * Interactive CLI interface
 */
class InteractiveCLI {
    constructor(client) {
        this.client = client;
        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            prompt: 'mcp> '
        });
    }

    /**
     * Start interactive mode
     */
    async start() {
        console.log('RAG MCP Server - Interactive Node.js Client');
        console.log('Type "help" for available commands, "quit" to exit');
        console.log();

        this.rl.prompt();

        this.rl.on('line', async (input) => {
            const trimmed = input.trim();
            if (!trimmed) {
                this.rl.prompt();
                return;
            }

            const [command, ...args] = trimmed.split(' ');

            try {
                await this.handleCommand(command.toLowerCase(), args);
            } catch (error) {
                console.error(`Error: ${error.message}`);
            }

            console.log();
            this.rl.prompt();
        });

        this.rl.on('close', () => {
            console.log('Goodbye!');
            process.exit(0);
        });
    }

    /**
     * Handle CLI commands
     */
    async handleCommand(command, args) {
        switch (command) {
            case 'help':
            case 'h':
                this.showHelp();
                break;

            case 'test':
            case 't':
                await this.client.testConnection();
                break;

            case 'status':
            case 's':
                const status = await this.client.getServerStatus();
                console.log(JSON.stringify(status, null, 2));
                break;

            case 'health':
                const health = await this.client.getHealth();
                console.log(JSON.stringify(health, null, 2));
                break;

            case 'ingest':
            case 'i':
                if (args.length === 0) {
                    console.log('Usage: ingest <file_path>');
                    break;
                }
                const filePath = args.join(' ');
                console.log(`Ingesting CSV file: ${filePath}`);
                const ingestResult = await this.client.ingestCsv(filePath);
                console.log(JSON.stringify(ingestResult, null, 2));
                break;

            case 'query':
            case 'q':
                if (args.length === 0) {
                    console.log('Usage: query <query_text> [max_results] [score_threshold]');
                    break;
                }
                
                const queryText = args.join(' ').replace(/\s+\d+(\.\d+)?$/, '').replace(/\s+\d+$/, '');
                const maxResults = parseInt(args[args.length - 2]) || 10;
                const scoreThreshold = parseFloat(args[args.length - 1]) || null;
                
                console.log(`Querying documents: ${queryText}`);
                const queryResults = await this.client.queryDocuments(queryText, maxResults, scoreThreshold);
                console.log(JSON.stringify(queryResults, null, 2));
                break;

            case 'quit':
            case 'exit':
            case 'q':
                this.rl.close();
                break;

            default:
                console.log(`Unknown command: ${command}`);
                console.log('Type "help" for available commands');
        }
    }

    /**
     * Show help information
     */
    showHelp() {
        console.log('Available commands:');
        console.log('  test                              - Test connection to server');
        console.log('  status                            - Get server status');
        console.log('  health                            - Get health status');
        console.log('  ingest <file_path>                - Ingest CSV file');
        console.log('  query <text> [max] [threshold]    - Query documents');
        console.log('  help                              - Show this help');
        console.log('  quit                              - Exit');
    }
}

/**
 * Command-line interface for non-interactive usage
 */
async function runCommand(client, command, args) {
    switch (command) {
        case 'test':
            const success = await client.testConnection();
            process.exit(success ? 0 : 1);
            break;

        case 'status':
            const status = await client.getServerStatus();
            console.log(JSON.stringify(status, null, 2));
            break;

        case 'health':
            const health = await client.getHealth();
            console.log(JSON.stringify(health, null, 2));
            break;

        case 'ingest':
            if (args.length === 0) {
                console.error('Usage: node simple-client.js ingest <file_path>');
                process.exit(1);
            }
            const filePath = args.join(' ');
            console.log(`Ingesting CSV file: ${filePath}`);
            const ingestResult = await client.ingestCsv(filePath);
            console.log(JSON.stringify(ingestResult, null, 2));
            break;

        case 'query':
            if (args.length === 0) {
                console.error('Usage: node simple-client.js query <query_text> [max_results] [score_threshold]');
                process.exit(1);
            }
            
            // Parse arguments - last two might be numbers
            let queryText = args.join(' ');
            let maxResults = 10;
            let scoreThreshold = null;
            
            // Check if last argument is a number (score threshold)
            const lastArg = args[args.length - 1];
            if (!isNaN(parseFloat(lastArg)) && isFinite(lastArg)) {
                scoreThreshold = parseFloat(lastArg);
                args.pop();
                
                // Check if second-to-last argument is also a number (max results)
                const secondLastArg = args[args.length - 1];
                if (!isNaN(parseInt(secondLastArg)) && isFinite(secondLastArg)) {
                    maxResults = parseInt(secondLastArg);
                    args.pop();
                }
                
                queryText = args.join(' ');
            } else {
                // Check if last argument is an integer (max results)
                if (!isNaN(parseInt(lastArg)) && isFinite(lastArg)) {
                    maxResults = parseInt(lastArg);
                    args.pop();
                    queryText = args.join(' ');
                }
            }
            
            console.log(`Querying documents: ${queryText}`);
            if (maxResults !== 10) console.log(`Max results: ${maxResults}`);
            if (scoreThreshold !== null) console.log(`Score threshold: ${scoreThreshold}`);
            
            const queryResults = await client.queryDocuments(queryText, maxResults, scoreThreshold);
            console.log(JSON.stringify(queryResults, null, 2));
            break;

        case 'interactive':
            const cli = new InteractiveCLI(client);
            await cli.start();
            break;

        default:
            console.error(`Unknown command: ${command}`);
            showUsage();
            process.exit(1);
    }
}

/**
 * Show usage information
 */
function showUsage() {
    console.log('RAG MCP Server - Node.js Client');
    console.log();
    console.log('Usage: node simple-client.js [server_url] [command] [arguments...]');
    console.log();
    console.log('Commands:');
    console.log('  test                              - Test connection to server');
    console.log('  status                            - Get server status');
    console.log('  health                            - Get health status');
    console.log('  ingest <file_path>                - Ingest CSV file');
    console.log('  query <text> [max] [threshold]    - Query documents');
    console.log('  interactive                       - Start interactive mode');
    console.log();
    console.log('Examples:');
    console.log('  node simple-client.js test');
    console.log('  node simple-client.js status');
    console.log('  node simple-client.js ingest ./data/sample_products.csv');
    console.log('  node simple-client.js query "What products are available?" 5 0.7');
    console.log('  node simple-client.js interactive');
    console.log('  node simple-client.js http://remote-server:8000 test');
}

/**
 * Main function
 */
async function main() {
    const args = process.argv.slice(2);
    
    // Check if axios is available
    try {
        require('axios');
    } catch (error) {
        console.error('Error: axios is required but not installed');
        console.error('Install it with: npm install axios');
        process.exit(1);
    }
    
    // Parse arguments
    let serverUrl = 'http://localhost:8000';
    let command = 'interactive';
    let commandArgs = [];
    
    if (args.length === 0) {
        // Default to interactive mode
        command = 'interactive';
    } else if (args[0].startsWith('http://') || args[0].startsWith('https://')) {
        // First argument is server URL
        serverUrl = args[0];
        command = args[1] || 'interactive';
        commandArgs = args.slice(2);
    } else if (args[0] === 'help' || args[0] === '--help' || args[0] === '-h') {
        showUsage();
        return;
    } else {
        // First argument is command
        command = args[0];
        commandArgs = args.slice(1);
    }
    
    console.log(`Connecting to RAG MCP Server at: ${serverUrl}`);
    
    // Create client
    const client = new RAGMCPClient(serverUrl);
    
    try {
        await runCommand(client, command, commandArgs);
    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

// Run main function if this script is executed directly
if (require.main === module) {
    main().catch(error => {
        console.error(`Fatal error: ${error.message}`);
        process.exit(1);
    });
}

module.exports = RAGMCPClient;